keps/sig-cloud-provider/20180530-cloud-controller-manager.md

```diff
@@ -321,13 +321,13 @@ the following subdirectories that contain the common features:
 
 * `dns` for DNS provider code.
 * `cni` for the Container Network Interface (CNI) driver.
-* `csi` for the Container Storage Interface (CSI) driver.
```

Comment:
I think everything else in this section (dns, csi, cni, flex and installer) should also be optional, thoughts @hogepodge @jagosan?

Reply:
yeah, I considered just doing that, but didn't want to overstep. =)

At a minimum, I see CNI and flex in that category. But I wasn't sure about installer and DNS.

Reply:
none of these should be in the scope of CCM. why will CNI and flex be in the scope of CCM?

we agreed on this 2 quarters back. CNI, CSI spec will be owned independently and implementations for a provider, will be provider sub-projects.

Reply:
> we agreed on this 2 quarters back. CNI, CSI spec will be owned independently and implementations for a provider, will be provider sub-projects.

I'm not sure this was ever formalized (correct me if I'm wrong though please), this KEP existed to document what we expect from the provider repos though. I think we agree here that the `k8s.io/cloud-provider-*` repos should focus on CCM with the _option_ to host other plugins like CSI, CNI, etc. Where external plugins like CSI and CNI live is not a standard that I think SIG Cloud Provider should impose though. 

> none of these should be in the scope of CCM. why will CNI and flex be in the scope of CCM?

CSI, CNI, etc aren't necessarily in scope of CCM but they work hand-in-hand so it may make sense to bundle them together.

Reply:
Also, this KEP seems outdated given it was created during the earlier days of KEPs. I'm not sure if bundling repo organization requirements into the same KEP as the CCM makes sense. Maybe a discussion for the mailing list https://groups.google.com/forum/#!topic/kubernetes-sig-cloud-provider/3oi-bbPX96E

Reply:
> I'm not sure this was ever formalized (correct me if I'm wrong though please), this KEP existed to document what we expect from the provider repos though. I think we agree here that the k8s.io/cloud-provider-* repos should focus on CCM with the option to host other plugins like CSI, CNI, etc. Where external plugins like CSI and CNI live is not a standard that I think SIG Cloud Provider should impose though.

As discussed in slack, this was verbally discussed and agreed upon in our technical planning sessions with @cheftako on Thursday to move CCM out-of-tree. 
We did not need to create anything official as there is a migration KEP for in-tree provider storage calls to map completely to CSI being led by a specific team. This migration is based on the logic that CCM out-of-tree will remove all in-tree storage calls.

AWS and GCP Providers today (potentially collapsing into SIG CP) host CSI drivers. Therefore hosting these drivers should be a provider choice as we maintain the drivers and yes the logic can be extended by SIG CP to other providers as a preferred standard.

> CSI, CNI, etc aren't necessarily in scope of CCM but they work hand-in-hand so it may make sense to bundle them together.

What do you mean by bundled together? 

> We had agreement in the last SIG call though that dns, cni, csi, flex and installer can be optional directories for cloud provider repos. Longer term I would like to see this KEP be split up though.

We agreed to hosting flat sub-projects coming from the providers to SIG CP for now - why add optional repositories? I prefer Aaron's suggestion of placing all provider specific plugins including CCM into a provider repository if the goal is proper organization. If I am not mistaken this is also a suggestion from @cheftako

Reply:
> As discussed in slack, this was verbally discussed and agreed upon in our technical planning sessions with @cheftako on Thursday to move CCM out-of-tree.

Yup, don't disagree that CCM code should be out-of-tree in external repos, this PR is only stating that CSI implementations don't have to be co-located. 

> AWS and GCP Providers today (potentially collapsing into SIG CP) host CSI drivers. Therefore hosting these drivers should be a provider choice as we maintain the drivers and yes the logic can be extended by SIG CP to other providers as a preferred standard.

Maybe I'm misunderstanding, but I think this is what the PR is proposing -  that the location of the CSI driver does not have to be bundled with the cloud provider repos.

Reply:
> What do you mean by bundled together?

CSI & CCM developed in the same repo, released together. Still different components in a cluster though 

> We agreed to hosting flat sub-projects coming from the providers to SIG CP for now - why add optional repositories? I prefer Aaron's suggestion of placing all provider specific plugins including CCM into a provider repository if the goal is proper organization. If I am not mistaken this is also a suggestion from @cheftako

There are benefits to hosting something like CSI in separate repositories. For example, [GCE PD](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver) and [AWS EBS](https://github.com/kubernetes-sigs/aws-ebs-csi-driver) CSI drivers live in separate repos today. This would also align better with the flat sub-project structure you mentioned since we can have CSI drivers categorized as it's own subproject. This PR is clarifying that plugins like CSI drivers be optionally hosted in external cloud provider repos since previously we had documented that external provider repos should host everything. Providers can still host CSI in the same repo as CCM if they want to.